File: Python/A_Project/B_Snake_and_Food/Utility/snake.py
from collections import deque

from Utility.pair import Pair


class Snake:
    def __init__(self, start=None):
        self.body_segments = deque()
        if start is None:
            self.body_segments.appendleft(Pair(0, 0))
        else:
            self.body_segments.appendleft(start)

    def head(self):
        return self.body_segments[0]

    def tail(self):
        return self.body_segments[-1]

    def move_to(self, new_head):
        self.body_segments.appendleft(new_head)
        self.body_segments.pop()

    def grow_to(self, new_head):
        self.body_segments.appendleft(new_head)

    def length(self) -> int:
        return len(self.body_segments)

    def body(self):
        return self.body_segments

File: Ruby/A_Project/B_Snake_And_Food/Utility/snake.rb
# frozen_string_literal: true

require_relative "pair"

class Snake
  def initialize(start = nil)
    @body_segments = []
    if start.nil?
      @body_segments.unshift(Pair.new(0, 0))
    else
      @body_segments.unshift(start)
    end
  end

  def head
    @body_segments[0]
  end

  def tail
    @body_segments[-1]
  end

  def move_to(new_head)
    @body_segments.unshift(new_head)
    @body_segments.pop
  end

  def grow_to(new_head)
    @body_segments.unshift(new_head)
  end

  def length
    @body_segments.length
  end

  def body
    @body_segments.dup
  end
end
